### PR TITLE
feat(ecr): add option to auto delete images upon ECR repository removal

### PR DESCRIPTION
**Pull request by randyridgley**
This request fixes the ECR Repository resource to allow setting a flag on the resource to auto delete the images in the repository. This is similar to the way S3 handles the autoDeleteObjects attribute. This code base starts from a stalled PR [#15932](https://github.com/aws/aws-cdk/pull/15932). This also takes into account the functionality added into S3 to create tag to not delete images if the flag is flipped from true to false. 

Closes [#12618](https://github.com/aws/aws-cdk/issues/12618)
References closed and not merged PR  [#15932](https://github.com/aws/aws-cdk/pull/15932)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by evgenyka** [comment]
@randyridgley please see https://github.com/aws/aws-cdk/pull/15932 for the same issue, which been already progressed.

**Review by MrArnoldPalmer — CHANGES_REQUESTED** [review]
Just some small feedback, otherwise this looks pretty good.

File: packages/@aws-cdk/aws-ecr/lib/auto-delete-images-handler/index.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { ECR } from 'aws-sdk';
+
+const AUTO_DELETE_IMAGES_TAG = 'aws-cdk:auto-delete-images';
+
+const ecr = new ECR();
+
+export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent) {
+  switch (event.RequestType) {
+    case 'Create':
```
[the remark this answers is not in the document]
Comment (randyridgley) [reply packages/@aws-cdk/aws-ecr/lib/auto-delete-images-handler/index.ts:10]: fixed by adding a break

File: packages/@aws-cdk/aws-ecr/lib/repository.ts [hunk 2]
```diff
@@ -741,6 +768,50 @@ export class Repository extends RepositoryBase {
...
+      runtime: CustomResourceProviderRuntime.NODEJS_14_X,
+      description: `Lambda function for auto-deleting images in ${this.repositoryName} repository.`,
+      policyStatements: [
+        {
+          Effect: 'Allow',
+          Action: [
+            'ecr:BatchDeleteImage',
+            'ecr:ListImages',
+          ],
+          Resource: ['*'], // TODO?
```
[the remark this answers is not in the document]
Comment (randyridgley) [reply packages/@aws-cdk/aws-ecr/lib/repository.ts]: merged both and pulled through the this._resource arn ref from the creation of the L1 construct

File: packages/@aws-cdk/aws-ecr/lib/repository.ts [hunk 3]
```diff
@@ -741,6 +768,50 @@ export class Repository extends RepositoryBase {
...
+        },
+        {
+          Effect: 'Allow',
+          Action: [
+            'ecr:DescribeRepositories',
+            'ecr:ListTagsForResource',
+          ],
+          Resource: ['*'],
```
[the remark this answers is not in the document]
Comment (randyridgley) [reply packages/@aws-cdk/aws-ecr/lib/repository.ts]: merged both and pulled through the this._resource arn ref from the creation of the L1 construct

File: packages/@aws-cdk/aws-ecr/lib/auto-delete-images-handler/index.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    case 'Delete':
+      return onDelete(event.ResourceProperties?.RepositoryName);
+  }
+}
+
+async function onUpdate(event: AWSLambda.CloudFormationCustomResourceEvent) {
+  const updateEvent = event as AWSLambda.CloudFormationCustomResourceUpdateEvent;
+  const oldRepositoryName = updateEvent.OldResourceProperties?.RepositoryName;
+  const newRepositoryName = updateEvent.ResourceProperties?.RepositoryName;
+  const repositoryNameHasChanged = newRepositoryName != null && oldRepositoryName != null && newRepositoryName !== oldRepositoryName;
```
[the remark this answers is not in the document]
Comment (randyridgley) [reply packages/@aws-cdk/aws-ecr/lib/auto-delete-images-handler/index.ts]: Thanks for the pointer! Cleaned up the conditionals based on your note.

File: packages/@aws-cdk/aws-ecr/lib/repository.ts [hunk 5]
```diff
@@ -1,12 +1,20 @@
...
+import * as path from 'path';
 import * as events from '@aws-cdk/aws-events';
 import * as iam from '@aws-cdk/aws-iam';
 import * as kms from '@aws-cdk/aws-kms';
-import { ArnFormat, IResource, Lazy, RemovalPolicy, Resource, Stack, Tags, Token, TokenComparison } from '@aws-cdk/core';
+import {
+  ArnFormat, IResource, Lazy, RemovalPolicy, Resource, Stack,
+  Tags, Token, TokenComparison, CustomResource, CustomResourceProvider,
+  CustomResourceProviderRuntime,
+} from '@aws-cdk/core';
```
Comment (MrArnoldPalmer) [inline packages/@aws-cdk/aws-ecr/lib/repository.ts:10]:
```suggestion
import {
  ArnFormat,
  IResource,
  Lazy,
  RemovalPolicy,
  Resource,
  Stack,
  Tags,
  Token,
  TokenComparison,
  CustomResource,
  CustomResourceProvider,
  CustomResourceProviderRuntime,
} from '@aws-cdk/core';
```

Just by convention

**Review by MrArnoldPalmer — DISMISSED** [review]
Just a small convention change, otherwise I think we are good.

**Pull request merged**